Optimize `checked_ilog` and `pow` when `base` is a power of two

Kmeakin opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/147250)*
<!-- homu-ignore:end -->

Optimize `checked_ilog` and `pow` when the base is a power of two

Kmeakin committed:
optimize: `checked_ilog` when `base` is a power of two

if base == 2 ** k, then
log(base, n) == log(2, n) / k

Kmeakin committed:
optimize: `pow` when `base` is a power of two

if base == 2 ** k, then
   (2 ** k) ** n
== 2 ** (k * n)
== 1 << (k * n)

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18180633059/job/51755637337) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51755637337) [(plain)](https://github.com/rust-lang/rust/commit/e5ca8cdedf973c018c5ccc39e12041ccca0764b5/checks/51755637337/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
note: no errors encountered even though delayed bugs were created

note: those delayed bugs will now be shown as internal compiler errors

error: internal compiler error: alignment greater than 2^29 should be errored on elsewhere
##[error]   --> example/mini_core_hello_world.rs:359:12
    |
359 |     #[repr(align(8192))]
    |            ^^^^^^^^^^^
    |
---
         3: <rustc_errors::diagnostic::Diag>::emit_producing_error_guaranteed
         4: <rustc_errors::DiagCtxtHandle>::span_delayed_bug::<rustc_span::span_encoding::Span, &str>
         5: <rustc_passes::check_attr::CheckAttrVisitor>::check_attributes
         6: <rustc_passes::check_attr::CheckAttrVisitor as rustc_hir::intravisit::Visitor>::visit_item
         7: <rustc_middle::ty::context::TyCtxt>::hir_visit_item_likes_in_module::<rustc_passes::check_attr::CheckAttrVisitor>
         8: rustc_passes::check_attr::check_mod_attrs
         9: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::check_mod_attrs::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 0]>>
        10: <rustc_query_impl::query_impl::check_mod_attrs::dynamic_query::{closure#2} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, rustc_span::def_id::LocalModDefId)>>::call_once
        11: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with::<rustc_middle::ty::context::tls::enter_context<rustc_query_system::query::plumbing::try_load_from_disk_and_cache_in_memory<rustc_query_impl::DynamicConfig<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 0]>, rustc_query_system::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt>::{closure#0}, rustc_middle::query::erase::Erased<[u8; 0]>>::{closure#0}, rustc_middle::query::erase::Erased<[u8; 0]>>
        12: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::LocalModDefId, rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, false>
        13: rustc_query_impl::query_impl::check_mod_attrs::get_query_non_incr::__rust_end_short_backtrace
        14: <rustc_middle::ty::context::TyCtxt>::par_hir_for_each_module::<rustc_interface::passes::run_required_analyses::{closure#0}::{closure#1}::{closure#0}::{closure#2}::{closure#0}>::{closure#0}
        15: <rustc_data_structures::sync::parallel::ParallelGuard>::run::<(), rustc_interface::passes::run_required_analyses::{closure#0}::{closure#0}::{closure#1}>
        16: rustc_interface::passes::analysis
        17: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 0]>>
        18: <rustc_query_impl::query_impl::analysis::dynamic_query::{closure#2} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, ())>>::call_once
        19: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with::<rustc_middle::ty::context::tls::enter_context<rustc_query_system::query::plumbing::execute_job_non_incr<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::SingleCache<rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt>::{closure#0}, rustc_middle::query::erase::Erased<[u8; 0]>>::{closure#0}, rustc_middle::query::erase::Erased<[u8; 0]>>
        20: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::SingleCache<rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, false>
        21: rustc_query_impl::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
        22: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with::<rustc_middle::ty::context::tls::enter_context<<rustc_middle::ty::context::GlobalCtxt>::enter<rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>::{closure#1}, core::option::Option<rustc_interface::queries::Linker>>::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>
        23: <rustc_middle::ty::context::TyCtxt>::create_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}>
        24: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
        25: <alloc::boxed::Box<dyn for<'a> core::ops::function::FnOnce<(&'a rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &'a std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena<'a>>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}), Output = core::option::Option<rustc_interface::queries::Linker>>> as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once
        26: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
        27: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
        28: rustc_span::create_session_globals_then::<(), rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}>
        29: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
        30: <<std::thread::Builder>::spawn_unchecked_<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
---
warning: the ICE couldn't be written to `/checkout/compiler/rustc_codegen_cranelift/rustc-ice-2025-10-02T02_02_05-34620.txt`: Read-only file system (os error 30)

note: rustc 1.92.0-nightly (ba891baca 2025-10-02) running on aarch64-unknown-linux-gnu

note: compiler flags: -C panic=abort -Z panic-abort-tests -Z codegen-backend=cranelift -C symbol-mangling-version=v0 -Z randomize-layout -Z unstable-options -Z macro-backtrace -C split-debuginfo=off -C link-arg=-L/usr/lib/llvm-20/lib -C llvm-args=-import-instr-limit=10 -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C debuginfo=2 -C panic=abort -Z force-unstable-if-unmarked

query stack during panic:
end of query stack
"/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/dist/rustc-clif" "-Csymbol-mangling-version=v0" "-Zrandomize-layout" "-Zunstable-options" "--check-cfg=cfg(bootstrap)" "-Zmacro-backtrace" "-Csplit-debuginfo=off" "-Clink-arg=-L/usr/lib/llvm-20/lib" "-Cllvm-args=-import-instr-limit=10" "-Clink-args=-Wl,-z,origin" "-Clink-args=-Wl,-rpath,$ORIGIN/../lib" "-Alinker-messages" "-L" "crate=/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/build/example" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/build/example" "-Cdebuginfo=2" "--target" "aarch64-unknown-linux-gnu" "-Cpanic=abort" "--check-cfg=cfg(jit)" "example/mini_core_hello_world.rs" exited with status ExitStatus(unix_wait_status(25856))
Command `/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo run --target aarch64-unknown-linux-gnu -Zbinary-dep-depinfo -j 4 -Zroot-dir=/checkout --locked --color always --release --manifest-path /checkout/compiler/rustc_codegen_cranelift/build_system/Cargo.toml -- test --download-dir /checkout/obj/build/cg_clif_download --out-dir /checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif --no-unstable-features --use-backend cranelift --sysroot llvm --skip-test testsuite.extended_sysroot [workdir=/checkout/compiler/rustc_codegen_cranelift]` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/test.rs:3658:25
Executed at: src/bootstrap/src/core/build_steps/test.rs:3703:26

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `--stage 2 test --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Build completed unsuccessfully in 0:22:45
  local time: Thu Oct  2 02:02:05 UTC 2025
  network time: Thu, 02 Oct 2025 02:02:05 GMT
##[error]Process completed with exit code 1.
```

</details>

workingjubilee commented:
does this affect the codegen in practical circumstances?

Kmeakin commented on library/core/src/num/uint_macros.rs:
```diff
@@ -1573,7 +1573,15 @@ macro_rules! uint_impl {
+                if base.is_power_of_two() {
+                    let k = base.ilog2();
+                    return Some(try_opt!(self.checked_ilog2()) / k);
+                }
```
Thanks! Fixed

nikic commented on library/core/src/num/uint_macros.rs:
```diff
@@ -3310,6 +3315,16 @@ macro_rules! uint_impl {
+                //    (2 ** k) ** n
+                // == 2 ** (k * n)
+                // == 1 << (k * n)
+                if self.is_power_of_two() {
```
Isn't the is_val_statically_known() check on the wrong value here? You need a known base, not a known exponent?

Kmeakin commented:
@rustbot ready

scottmcm reviewed (changes requested):
The code here is looking good, but can you make sure there's normal runtime behaviour tests for it too?  Notably, after the base conversation (that's fixed in the code) it made me think that that's not currently covered by any tests, so we should have something -- maybe just add to the `# Examples` that it returns None for base zero or one?  (They seem like perfectly reasonable and helpful examples, in addition to giving coverage for this stuff.)

And maybe add some `should_panic` tests to ensure that the overflow checking is still correct for `pow` when overflow checks are enabled?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147250/bcbbdcb8522fd3cb4a8dde62313b251ab107694d..c84b99eefcb648b9e68a104103ea5169ced4faac)*

Kmeakin commented on tests/codegen-llvm/ilog_known_base.rs:
```diff
@@ -0,0 +1,36 @@
+#[no_mangle]
+pub fn checked_ilog16(val: u32) -> Option<u32> {
+    // CHECK: %[[ICMP:.+]] = icmp ne i32 %val, 0
+    // CHECK: %[[CTZ:.+]] = tail call range(i32 0, 33) i32 @llvm.ctlz.i32(i32 %val, i1 true)
```
The `range` attribute is too large, but it seems LLVM is still able to propagate the knowledge:

https://godbolt.org/z/dzxfGnhMf

scottmcm commented on library/core/src/num/uint_macros.rs:
```diff
@@ -3309,6 +3318,18 @@ macro_rules! uint_impl {
+                // == 1 << (k * n)
+                if base.is_power_of_two() {
+                    let k = base.ilog2();
+                    return 1 << (k * exp)
```
I've been staring at this line, and I think it's wrong in release mode.

Take `2_u16.pow(16)`, for example.  Today that gives zero, which is what I'd expect, since that's the infinite-precision result wrapped to `u16`, but the shift would be `1_u16 << (1 * 16)` ⇒ `1_u16 << 16` ⇒ `1_u16`.

So please add some more tests to either show I'm wrong or fix it if I'm right.  Be sure to cover things like `4_u16.pow(0x8000_0000)` too, since the multiplication of the exponent could also overflow.

Kmeakin replied:
I think you're right. We should use `1 << (k * exp)` when overflow checks are enabled, and `1.unbounded_shl(k * exp)` when overflow checks are disabled. But I don't think there is a way to choose between the two depending on the caller crate's compiler flags. There is [`cfg(overflow_checks)`](github.com/rust-lang/rust/issues/111466) but it sounds like it would always resolve to `false` in the `core` crate

scottmcm reviewed (changes requested):
I still have concerns about the correctness of this; see above.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147250/a41214f9bd7ce869d21898f31107f92c90a9112f..53449c990cee855ae726e5ce7a08f3f3148c66f7)*

Kmeakin committed:
refactor: Increase test coverage for `pow`

Increase test coverage to check all interesting edge cases and all
variants.

Kmeakin committed:
refactor: Deduplicate `pow` implementations

`strict_pow` can be implemented in terms of `checked_pow`,
`wrapping_pow` can be implemented in terms of `overflowing_pow`,
and `pow` can be implemented in terms of `strict_pow` or `wrapping_pow`.

Kmeakin committed:
optimize: `pow` if `exp` is statically known

Copy the optimization that unrolls the loop from `pow` to `checked_pow`
and `overflowing_pow`.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19614818872/job/56165891923) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/56165891923) [(plain)](https://github.com/rust-lang/rust/commit/c2f36c5b55b25314018c0789f8dfd7d4c527a246/checks/56165891923/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    num::u8::panicking_pow

test result: FAILED. 2253 passed; 5 failed; 6 ignored; 0 measured; 0 filtered out; finished in 2.99s

error: test failed, to rerun pass `-p coretests --test coretests`
env -u RUSTC_WRAPPER CARGO_ENCODED_RUSTDOCFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zannotate-moves\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}-Dwarnings\u{1f}-Wrustdoc::invalid_codeblock_attributes\u{1f}--crate-version\u{1f}1.93.0-nightly\t(739c484aa\t2025-11-23)" CARGO_ENCODED_RUSTFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zannotate-moves\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}-Zmacro-backtrace\u{1f}-Csplit-debuginfo=off\u{1f}-Clink-arg=-L/usr/lib/llvm-20/lib\u{1f}-Cllvm-args=-import-instr-limit=10\u{1f}-Clink-args=-Wl,-z,origin\u{1f}-Clink-args=-Wl,-rpath,$ORIGIN/../lib\u{1f}-Alinker-messages\u{1f}--cap-lints=allow\u{1f}--cfg\u{1f}randomized_layouts" RUSTC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/dist/rustc-clif" RUSTDOC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/dist/rustdoc-clif" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "test" "--manifest-path" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/build/sysroot_tests/Cargo.toml" "--target-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/build/sysroot_tests_target" "--locked" "--target" "aarch64-unknown-linux-gnu" "-p" "coretests" "-p" "alloctests" "--tests" "--" "-q" exited with status ExitStatus(unix_wait_status(25856))
Command `/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo run --target aarch64-unknown-linux-gnu -Zbinary-dep-depinfo -j 4 -Zroot-dir=/checkout --locked --color always --release --manifest-path /checkout/compiler/rustc_codegen_cranelift/build_system/Cargo.toml -- test --download-dir /checkout/obj/build/cg_clif_download --out-dir /checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif --no-unstable-features --use-backend cranelift --sysroot llvm --skip-test testsuite.extended_sysroot [workdir=/checkout/compiler/rustc_codegen_cranelift]` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/test.rs:3723:25
Executed at: src/bootstrap/src/core/build_steps/test.rs:3768:26

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `--stage 2 test --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Build completed unsuccessfully in 0:22:50
  local time: Sun Nov 23 18:01:23 UTC 2025
  network time: Sun, 23 Nov 2025 18:01:23 GMT
##[error]Process completed with exit code 1.
```

</details>

quaternic commented on library/core/src/num/uint_macros.rs:
```diff
@@ -3277,30 +3267,60 @@ macro_rules! uint_impl {
+                (acc, tmp_overflow) = acc.overflowing_mul(base);
+                overflow |= tmp_overflow;
+                return (acc, overflow);
+            }
```
So, here's an idea: If either input is statically known, checking for overflow could be folded into a single range check on the other input. Doing so would allow all of the variants to delegate to `wrapping_pow`.

Quick test that this would help:
https://rust.godbolt.org/z/oEWWGx4f9

quaternic replied:
Although, I suppose it's really the known exponent cases that would benefit the most by removing the need for having the LLVM-unrollable blocks duplicated for each variant.

In case it's of use, here's the code I tested with: https://godbolt.org/z/5svnKM7hM
(Includes computing the roots for unsigned types. Signed would have some additional complications)

Kmeakin replied:
I think the PR is getting complicated enough already. Let's get this merged first, then we can further optimise in a follow up

Kmeakin replied:
I think the PR is getting complicated enough already. Let's get this merged first, then we can further optimise in a follow up PR

Kmeakin commented:
@scottmcm ping?

scottmcm commented:
@rustbot reroll

jhpratt commented on library/core/src/num/int_macros.rs:
```diff
@@ -1761,11 +1761,41 @@ macro_rules! int_impl {
+            let mut acc: Self = 1;
+
+            if intrinsics::is_val_statically_known(base) {
+                if base.unsigned_abs().is_power_of_two() {
```
Is there a reason these two aren't combined? I wouldn't think that having them separate would affect how the intrinsic optimizes, but weirder things have happened. This is present in a couple places.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147250/f5eca4fcfa908d1e038afd19c6e746f075859130..f23f1e5affc540903782cbf235cf45d04eef2559)*

jhpratt commented:
With the one nit resolved (assuming there's no reason to leave as-is), LGTM.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

quaternic commented on library/core/src/num/int_macros.rs:
```diff
@@ -1845,11 +1845,39 @@ macro_rules! int_impl {
+            if intrinsics::is_val_statically_known(base) && base.unsigned_abs().is_power_of_two() {
+                let k = base.unsigned_abs().ilog2();
+                let shift = try_opt!(k.checked_mul(exp));
+                let magnitude = try_opt!((1 as Self).checked_shl(shift));
```
If `shift == Self::BITS - 1`, this gives `magnitude = Self::MIN`, and returns `Some(Self::MIN)` instead of `None` for e.g. `1_i32.checked_pow(31)`. `overflowing_pow` has the same issue.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147250/93637f398f4d0d40badc6bf1ed3f241499d06f88..f20dfa88d46fae42f178543a7366f426c25c5a62)*

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-21-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25196236737/job/73877473092) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/73877473092) [(plain)](https://github.com/rust-lang/rust/commit/6986ad2c495e39cd4133ae07c86cb329a7a6ba81/checks/73877473092/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
Executing "/scripts/stage_2_test_set1.sh"
+ /scripts/stage_2_test_set1.sh
+ '[' 1 == 1 ']'
+ echo 'PR_CI_JOB set; skipping tidy'
+ SKIP_TIDY='--skip tidy'
+ ../x.py --stage 2 test --skip tidy --skip compiler --skip src
PR_CI_JOB set; skipping tidy
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
---
test [codegen] tests/codegen-llvm/zip.rs ... ok

failures:

---- [codegen] tests/codegen-llvm/pow_known_base.rs stdout ----
------FileCheck stdout------------------------------

------FileCheck stderr------------------------------
/checkout/tests/codegen-llvm/pow_known_base.rs:51:12: error: CHECK: expected string not found in input
 // CHECK: %[[POW:.+]] = shl nuw i32 1, %exp
           ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll:47:32: note: scanning from here
 %_71.i = icmp ult i32 %exp, 32
                               ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll:54:2: note: possible intended match here
 %1 = shl nsw i32 -1, %exp
 ^

Input file: /checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll
Check file: /checkout/tests/codegen-llvm/pow_known_base.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'pow_known_base.98717cbbfaeb9c42-cgu.0' 
            2: source_filename = "pow_known_base.98717cbbfaeb9c42-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32" 
            4: target triple = "aarch64-unknown-linux-gnu" 
            5:  
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
            7: define noundef range(i32 0, -2147483647) i32 @pow2(i32 noundef %exp) unnamed_addr #0 { 
            8: start: 
            9:  %_45.i = icmp ult i32 %exp, 32 
           10:  %0 = shl nuw i32 1, %exp 
           11:  %_9.sroa.0.0.i = select i1 %_45.i, i32 %0, i32 0 
           12:  ret i32 %_9.sroa.0.0.i 
           13: } 
           14:  
           15: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           16: define noundef range(i32 0, -2147483647) i32 @pow4(i32 noundef %exp) unnamed_addr #0 { 
           17: start: 
           18:  %_43.1.i = icmp slt i32 %exp, 0 
           19:  %_43.0.i = shl i32 %exp, 1 
           20:  %_45.i = icmp ult i32 %_43.0.i, 32 
           21:  %0 = shl nuw i32 1, %_43.0.i 
           22:  %_9.sroa.0.0.i = select i1 %_45.i, i32 %0, i32 0 
           23:  %_0.sroa.0.0.i = select i1 %_43.1.i, i32 0, i32 %_9.sroa.0.0.i, !prof !2 
           24:  ret i32 %_0.sroa.0.0.i 
           25: } 
           26:  
           27: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           28: define noundef range(i32 0, -2147483647) i32 @pow16(i32 noundef %exp) unnamed_addr #0 { 
           29: start: 
           30:  %_43.1.i = icmp ugt i32 %exp, 1073741823 
           31:  %_43.0.i = shl i32 %exp, 2 
           32:  %_45.i = icmp ult i32 %_43.0.i, 32 
           33:  %0 = shl nuw i32 1, %_43.0.i 
           34:  %_9.sroa.0.0.i = select i1 %_45.i, i32 %0, i32 0 
           35:  %_0.sroa.0.0.i = select i1 %_43.1.i, i32 0, i32 %_9.sroa.0.0.i, !prof !2 
           36:  ret i32 %_0.sroa.0.0.i 
           37: } 
           38:  
           39: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           40: define noundef range(i32 -2147483648, 1073741825) i32 @pow_minus_2(i32 noundef %exp) unnamed_addr #0 { 
           41: start: 
           42:  %_14.i = and i32 %exp, 1 
           43:  %.not.i = icmp eq i32 %_14.i, 0 
           44:  br i1 %.not.i, label %bb8.i, label %bb5.i 
           45:  
           46: bb8.i: ; preds = %start 
           47:  %_71.i = icmp ult i32 %exp, 32 
check:51'0                                    X error: no match found
           48:  %0 = shl nuw nsw i32 1, %exp 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49:  %_19.sroa.0.0.i = select i1 %_71.i, i32 %0, i32 0 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50:  br label %_RNvMs0_NtCsjKeDKyqFnhH_4core3numl15overflowing_pow.exit 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51:  
check:51'0     ~
           52: bb5.i: ; preds = %start 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           53:  %_61.i = icmp ugt i32 %exp, 31 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           54:  %1 = shl nsw i32 -1, %exp 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:51'1      ?                          possible intended match
           55:  %_15.sroa.0.0.i = select i1 %_61.i, i32 0, i32 %1 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           56:  br label %_RNvMs0_NtCsjKeDKyqFnhH_4core3numl15overflowing_pow.exit 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           57:  
check:51'0     ~
           58: _RNvMs0_NtCsjKeDKyqFnhH_4core3numl15overflowing_pow.exit: ; preds = %bb8.i, %bb5.i 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           59:  %_0.sroa.0.0.i = phi i32 [ %_15.sroa.0.0.i, %bb5.i ], [ %_19.sroa.0.0.i, %bb8.i ] 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           60:  ret i32 %_0.sroa.0.0.i 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           61: } 
check:51'0     ~~
           62:  
check:51'0     ~
           63: attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable "frame-pointer"="non-leaf" "probe-stack"="inline-asm" "target-cpu"="generic" "target-features"="+v8a,+outline-atomics" } 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           64:  
check:51'0     ~
           65: !llvm.module.flags = !{!0} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           66: !llvm.ident = !{!1} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~
           67:  
check:51'0     ~
           68: !0 = !{i32 8, !"PIC Level", i32 2} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           69: !1 = !{!"rustc version 1.97.0-nightly (a8080f491 2026-05-01)"} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           70: !2 = !{!"branch_weights", !"expected", i32 1, i32 2000} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>

------------------------------------------

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-21/bin/FileCheck" "--input-file" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll" "/checkout/tests/codegen-llvm/pow_known_base.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen-llvm/pow_known_base.rs:51:12: error: CHECK: expected string not found in input
 // CHECK: %[[POW:.+]] = shl nuw i32 1, %exp
           ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll:47:32: note: scanning from here
 %_71.i = icmp ult i32 %exp, 32
                               ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll:54:2: note: possible intended match here
 %1 = shl nsw i32 -1, %exp
 ^

Input file: /checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll
Check file: /checkout/tests/codegen-llvm/pow_known_base.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'pow_known_base.98717cbbfaeb9c42-cgu.0' 
            2: source_filename = "pow_known_base.98717cbbfaeb9c42-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32" 
            4: target triple = "aarch64-unknown-linux-gnu" 
            5:  
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
            7: define noundef range(i32 0, -2147483647) i32 @pow2(i32 noundef %exp) unnamed_addr #0 { 
            8: start: 
            9:  %_45.i = icmp ult i32 %exp, 32 
           10:  %0 = shl nuw i32 1, %exp 
           11:  %_9.sroa.0.0.i = select i1 %_45.i, i32 %0, i32 0 
           12:  ret i32 %_9.sroa.0.0.i 
           13: } 
           14:  
           15: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           16: define noundef range(i32 0, -2147483647) i32 @pow4(i32 noundef %exp) unnamed_addr #0 { 
           17: start: 
           18:  %_43.1.i = icmp slt i32 %exp, 0 
           19:  %_43.0.i = shl i32 %exp, 1 
           20:  %_45.i = icmp ult i32 %_43.0.i, 32 
           21:  %0 = shl nuw i32 1, %_43.0.i 
           22:  %_9.sroa.0.0.i = select i1 %_45.i, i32 %0, i32 0 
           23:  %_0.sroa.0.0.i = select i1 %_43.1.i, i32 0, i32 %_9.sroa.0.0.i, !prof !2 
           24:  ret i32 %_0.sroa.0.0.i 
           25: } 
           26:  
           27: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           28: define noundef range(i32 0, -2147483647) i32 @pow16(i32 noundef %exp) unnamed_addr #0 { 
           29: start: 
           30:  %_43.1.i = icmp ugt i32 %exp, 1073741823 
           31:  %_43.0.i = shl i32 %exp, 2 
           32:  %_45.i = icmp ult i32 %_43.0.i, 32 
           33:  %0 = shl nuw i32 1, %_43.0.i 
           34:  %_9.sroa.0.0.i = select i1 %_45.i, i32 %0, i32 0 
           35:  %_0.sroa.0.0.i = select i1 %_43.1.i, i32 0, i32 %_9.sroa.0.0.i, !prof !2 
           36:  ret i32 %_0.sroa.0.0.i 
           37: } 
           38:  
           39: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           40: define noundef range(i32 -2147483648, 1073741825) i32 @pow_minus_2(i32 noundef %exp) unnamed_addr #0 { 
           41: start: 
           42:  %_14.i = and i32 %exp, 1 
           43:  %.not.i = icmp eq i32 %_14.i, 0 
           44:  br i1 %.not.i, label %bb8.i, label %bb5.i 
           45:  
           46: bb8.i: ; preds = %start 
           47:  %_71.i = icmp ult i32 %exp, 32 
check:51'0                                    X error: no match found
           48:  %0 = shl nuw nsw i32 1, %exp 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49:  %_19.sroa.0.0.i = select i1 %_71.i, i32 %0, i32 0 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50:  br label %_RNvMs0_NtCsjKeDKyqFnhH_4core3numl15overflowing_pow.exit 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51:  
check:51'0     ~
           52: bb5.i: ; preds = %start 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           53:  %_61.i = icmp ugt i32 %exp, 31 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           54:  %1 = shl nsw i32 -1, %exp 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:51'1      ?                          possible intended match
           55:  %_15.sroa.0.0.i = select i1 %_61.i, i32 0, i32 %1 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           56:  br label %_RNvMs0_NtCsjKeDKyqFnhH_4core3numl15overflowing_pow.exit 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           57:  
check:51'0     ~
           58: _RNvMs0_NtCsjKeDKyqFnhH_4core3numl15overflowing_pow.exit: ; preds = %bb8.i, %bb5.i 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           59:  %_0.sroa.0.0.i = phi i32 [ %_15.sroa.0.0.i, %bb5.i ], [ %_19.sroa.0.0.i, %bb8.i ] 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           60:  ret i32 %_0.sroa.0.0.i 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           61: } 
check:51'0     ~~
           62:  
check:51'0     ~
           63: attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable "frame-pointer"="non-leaf" "probe-stack"="inline-asm" "target-cpu"="generic" "target-features"="+v8a,+outline-atomics" } 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           64:  
check:51'0     ~
           65: !llvm.module.flags = !{!0} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           66: !llvm.ident = !{!1} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~
           67:  
check:51'0     ~
           68: !0 = !{i32 8, !"PIC Level", i32 2} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           69: !1 = !{!"rustc version 1.97.0-nightly (a8080f491 2026-05-01)"} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           70: !2 = !{!"branch_weights", !"expected", i32 1, i32 2000} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>
------------------------------------------

---- [codegen] tests/codegen-llvm/pow_known_base.rs stdout end ----

failures:
    [codegen] tests/codegen-llvm/pow_known_base.rs

test result: FAILED. 874 passed; 1 failed; 184 ignored; 0 measured; 3 filtered out; finished in 18.63s
```

</details>

Kmeakin commented:
ping @jhpratt?

jhpratt commented:
The last email I had received was the CI failure. I'll take a look when I can, but I'm busy in Europe at the moment.

jhpratt commented:
@bors try @rust-timer queue

rust-timer commented:
This pull request is already queued and waiting for a try build to finish.

<!-- rust-timer: temporary -->

rust-timer commented:
Queued c0638d923bca5c150f08ba8dd83d5ac9020f5095 with parent c58275e0369d09fc3959b8ba87dcbcbe73797465, future [comparison URL](https://perf.rust-lang.org/compare.html?start=c58275e0369d09fc3959b8ba87dcbcbe73797465&end=c0638d923bca5c150f08ba8dd83d5ac9020f5095).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.0 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (c0638d923bca5c150f08ba8dd83d5ac9020f5095): [comparison URL](https://perf.rust-lang.org/compare.html?start=c58275e0369d09fc3959b8ba87dcbcbe73797465&end=c0638d923bca5c150f08ba8dd83d5ac9020f5095&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read:


Benchmarking means the PR may be perf-sensitive. It's automatically marked not fit for rolling up. Overriding is possible but disadvised: it risks changing compiler perf.

**Next, please**: If you can, justify the regressions found in this try perf run in writing along with `@rustbot label: +perf-regression-triaged`. If not, fix the regressions and do another perf run. Neutral or positive results will clear the label automatically.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=c58275e0369d09fc3959b8ba87dcbcbe73797465&end=c0638d923bca5c150f08ba8dd83d5ac9020f5095&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.4%, 0.9%]   | 3     |
| Regressions ❌ <br /> (secondary)  | 0.4%  | [0.2%, 0.6%]   | 2     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.1%, -0.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -0.1% | [-0.1%, -0.1%] | 1     |
| All ❌✅ (primary)                 | 0.4%  | [-0.1%, 0.9%]  | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=c58275e0369d09fc3959b8ba87dcbcbe73797465&end=c0638d923bca5c150f08ba8dd83d5ac9020f5095&stat=max-rss)
<details>
<summary>Results (primary 2.8%, secondary 1.6%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.8% | [2.8%, 2.8%] | 1     |
| Regressions ❌ <br /> (secondary)  | 1.6% | [1.6%, 1.6%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.8% | [2.8%, 2.8%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=c58275e0369d09fc3959b8ba87dcbcbe73797465&end=c0638d923bca5c150f08ba8dd83d5ac9020f5095&stat=cycles:u)
<details>
<summary>Results (primary 2.3%, secondary 2.4%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.3% | [2.3%, 2.3%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.4% | [2.4%, 2.4%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.3% | [2.3%, 2.3%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=c58275e0369d09fc3959b8ba87dcbcbe73797465&end=c0638d923bca5c150f08ba8dd83d5ac9020f5095&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%, secondary 0.3%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.3%, 0.3%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.3%, 0.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.4%, -0.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.0% | [-0.4%, 0.3%]  | 2     |
</details>

**Bootstrap**: 515.367s -> 512.478s (-0.56%)
**Artifact size**: 400.82 MiB -> 400.82 MiB (0.00%)

jhpratt commented:
Hmm...those regressions are far beyond what would be considered significant (both statistically and by looking at the graphs). Any thoughts?

Kmeakin commented:
I'm baffled, honestly. As far as I can tell, these functions are only called by rustc at `const` time, or with variable bases, so the optimzation shouldn't change the runtime performance

* `ilog`
  * in `library/core/src/fmt/num_buffer.rs`: called at `const` time, and base is not a power of 2, so shouldn't affect runtime performance
  * in `library/core/src/fmt/num.rs`: base is a power of two (either 2, 8 or 16), but called at `const` time so shouldn't affect runtime performance
* `pow`
  * in `library/core/src/fmt/num.rs`: the base is 10, so shouldn't affect runtime performance
  * in `library/core/src/num/imp/int_sqrt.rs`: called at `const` time, so shouldn't affect runtime performance
  * in `library/std/src/sync/mpmc/utils.rs`: called at runtime with a variable base, so shouldn't affect runtime performance

jhpratt commented:
I have no idea either. Let's try another perf run. If it comes back the same (and for the same crates), then there's some weird issue. If it's different crates, it's clearly spurious.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued 759376f9a4e9acc6447f3f4933a190e8a1bb0ba1 with parent 4f84d9fac456d973d592cf3fb48db958ecf22506, future [comparison URL](https://perf.rust-lang.org/compare.html?start=4f84d9fac456d973d592cf3fb48db958ecf22506&end=759376f9a4e9acc6447f3f4933a190e8a1bb0ba1).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.0 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (759376f9a4e9acc6447f3f4933a190e8a1bb0ba1): [comparison URL](https://perf.rust-lang.org/compare.html?start=4f84d9fac456d973d592cf3fb48db958ecf22506&end=759376f9a4e9acc6447f3f4933a190e8a1bb0ba1&stat=instructions:u).

### Overall result: ❌ regressions - please read:


Benchmarking means the PR may be perf-sensitive. It's automatically marked not fit for rolling up. Overriding is possible but disadvised: it risks changing compiler perf.

**Next, please**: If you can, justify the regressions found in this try perf run in writing along with `@rustbot label: +perf-regression-triaged`. If not, fix the regressions and do another perf run. Neutral or positive results will clear the label automatically.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=4f84d9fac456d973d592cf3fb48db958ecf22506&end=759376f9a4e9acc6447f3f4933a190e8a1bb0ba1&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6% | [0.4%, 0.8%] | 3     |
| Regressions ❌ <br /> (secondary)  | 0.2% | [0.2%, 0.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.6% | [0.4%, 0.8%] | 3     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4f84d9fac456d973d592cf3fb48db958ecf22506&end=759376f9a4e9acc6447f3f4933a190e8a1bb0ba1&stat=max-rss)
<details>
<summary>Results (primary 3.0%, secondary -1.3%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.0%  | [3.0%, 3.0%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.8%  | [0.9%, 4.8%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-3.0%, -0.7%] | 9     |
| All ❌✅ (primary)                 | 3.0%  | [3.0%, 3.0%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4f84d9fac456d973d592cf3fb48db958ecf22506&end=759376f9a4e9acc6447f3f4933a190e8a1bb0ba1&stat=cycles:u)
<details>
<summary>Results (secondary -1.6%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.6% | [-1.6%, -1.6%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=4f84d9fac456d973d592cf3fb48db958ecf22506&end=759376f9a4e9acc6447f3f4933a190e8a1bb0ba1&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%, secondary 0.3%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.3%, 0.3%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.3%, 0.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.4%, -0.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.0% | [-0.4%, 0.3%]  | 2     |
</details>

**Bootstrap**: 510.625s -> 510.701s (0.01%)
**Artifact size**: 400.75 MiB -> 400.68 MiB (-0.02%)

jhpratt commented:
Same crates and for roughly the same amount. I'm clueless. Maybe try posting on Zulip, as I'm not sure who to ping.

jhpratt commented:
Despite the unexpected perf result, it was recommended on Zulip to merge this anyway.

@bors r+

JonathanBrouwer commented:
@bors p=2
Putting this above the rollupable PRs to maximize throughput, since the queue is busy

JonathanBrouwer commented:
Stuck
@bors cancel

Kmeakin commented:
> Despite the unexpected perf result, it was recommended on Zulip to merge this anyway.
> 
> @bors r+

Thanks. I was going to post to zulip but haven't had time this week. Can you link to the zulip discussion?

jhpratt commented:
https://rust-lang.zulipchat.com/#narrow/channel/247081-t-compiler.2Fperformance/topic/sus.20perf.20results/near/600950235

rust-timer commented:
Finished benchmarking commit (f20a92ec01483dc5c58e90e246f266bdad822d86): [comparison URL](https://perf.rust-lang.org/compare.html?start=43a4909ee98ed4d006d9d773f5d94dc58e34f846&end=f20a92ec01483dc5c58e90e246f266bdad822d86&stat=instructions:u).

### Overall result: ❌ regressions - please read:


Our benchmarks found a performance regression caused by this PR.
This might be an actual regression, but it can also be just noise.

**Next Steps**:

- If the regression was expected or you think it can be justified,
please write a comment with sufficient written justification, and add
`@rustbot label: +perf-regression-triaged` to it, to mark the regression as triaged.
- If you think that you know of a way to resolve the regression, try to create
a new PR with a fix for the regression.
- If you do not understand the regression or you think that it is just noise,
you can ask the `@rust-lang/wg-compiler-performance` working group for help (members of this group
were already notified of this PR).

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=43a4909ee98ed4d006d9d773f5d94dc58e34f846&end=f20a92ec01483dc5c58e90e246f266bdad822d86&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6% | [0.3%, 0.7%] | 4     |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.4%, 0.4%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.6% | [0.3%, 0.7%] | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=43a4909ee98ed4d006d9d773f5d94dc58e34f846&end=f20a92ec01483dc5c58e90e246f266bdad822d86&stat=max-rss)
<details>
<summary>Results (primary 5.0%, secondary -5.5%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 5.0%  | [5.0%, 5.0%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -5.5% | [-5.5%, -5.5%] | 1     |
| All ❌✅ (primary)                 | 5.0%  | [5.0%, 5.0%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=43a4909ee98ed4d006d9d773f5d94dc58e34f846&end=f20a92ec01483dc5c58e90e246f266bdad822d86&stat=cycles:u)
This perf run didn't have relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=43a4909ee98ed4d006d9d773f5d94dc58e34f846&end=f20a92ec01483dc5c58e90e246f266bdad822d86&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%, secondary 0.2%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.1%, 0.3%]   | 4     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.3%, 0.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.4%, -0.0%] | 6     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 1     |
| All ❌✅ (primary)                 | -0.0% | [-0.4%, 0.3%]  | 10    |
</details>

**Bootstrap**: 516.093s -> 516.04s (-0.01%)
**Artifact size**: 400.83 MiB -> 401.28 MiB (0.11%)

jhpratt commented:
This was known before approving.

@rustbot label +perf-regression-triaged
